File: src/test/java/de/fraunhofer/aisec/codyze/crymlin/AnalysisServerQueriesTest.java

package de.fraunhofer.aisec.codyze.crymlin;

import de.fraunhofer.aisec.codyze.analysis.cpgpasses.EdgeCachePass;
import de.fraunhofer.aisec.codyze.analysis.cpgpasses.IdentifierPass;
import de.fraunhofer.aisec.codyze.analysis.server.AnalysisServer;
import de.fraunhofer.aisec.codyze.analysis.structures.AnalysisContext;
import de.fraunhofer.aisec.codyze.analysis.structures.ServerConfiguration;
import de.fraunhofer.aisec.cpg.TranslationConfiguration;
import de.fraunhofer.aisec.cpg.TranslationManager;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.net.URL;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;

class AnalysisServerQueriesTest {

	private static AnalysisServer server;
	private static AnalysisContext result;

	@BeforeAll
	public static void startup() throws Exception {
		ClassLoader classLoader = AnalysisServerQueriesTest.class.getClassLoader();

		URL resource = classLoader.getResource("good/Bouncycastle.java");
		assertNotNull(resource);
		File javaFile = new File(resource.getFile());
		assertNotNull(javaFile);

		resource = classLoader.getResource("mark/PoC_MS1/Botan_AutoSeededRNG.mark");
		assertNotNull(resource);
		File markPoC1 = new File(resource.getFile());
		assertNotNull(markPoC1);
		String markModelFiles = markPoC1.getParent();

		// Start an analysis server
		server = AnalysisServer.builder()
				.config(
					ServerConfiguration.builder().launchConsole(false).launchLsp(false).markFiles(markModelFiles).build())
				.build();
		server.start();

		// Start the analysis
		TranslationManager translationManager = newJavaAnalysisRun(javaFile);
		CompletableFuture<AnalysisContext> analyze = server.analyze(translationManager);
		try {
			result = analyze.get(5, TimeUnit.MINUTES);
		}
		catch (TimeoutException t) {
			analyze.cancel(true);
			throw t;
		}
	}

	@AfterAll
	public static void teardown() {
		// Stop the analysis server
		server.stop();
	}

	/**
	 * Test analysis context - additional in-memory structures used for analysis.
	 */
	@Test
	void contextTest() {
		// Get analysis context from scratch
		AnalysisContext ctx = AnalysisServerQueriesTest.result;

		assertNotNull(ctx);
	}

	/**
	 * Helper method for initializing an Analysis Run.
	 *
	 * @param sourceLocations
	 * @return
	 */
	private static TranslationManager newJavaAnalysisRun(File... sourceLocations) {
		return TranslationManager.builder()
				.config(
					TranslationConfiguration.builder()
							.debugParser(true)
							.failOnError(false)
							.defaultPasses()
							.defaultLanguages()
							.registerPass(new IdentifierPass())
							.registerPass(new EdgeCachePass())
							.sourceLocations(
								sourceLocations)
							.build())
				.build();
	}
}

File: src/test/java/de/fraunhofer/aisec/codyze/crymlin/AnalysisServerQueriesTest.kt
package de.fraunhofer.aisec.codyze.crymlin

import de.fraunhofer.aisec.codyze.analysis.AnalysisContext
import de.fraunhofer.aisec.codyze.analysis.AnalysisServer
import de.fraunhofer.aisec.codyze.config.CodyzeConfiguration
import de.fraunhofer.aisec.codyze.config.Configuration
import de.fraunhofer.aisec.codyze.config.CpgConfiguration
import java.io.*
import java.lang.Exception
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import kotlin.Throws
import kotlin.test.assertNotNull
import org.junit.jupiter.api.*

internal class AnalysisServerQueriesTest : AbstractTest() {
    /** Test analysis context - additional in-memory structures used for analysis. */
    @Test
    fun contextTest() {
        // Get analysis context from scratch
        val ctx = result
        assertNotNull(ctx)
    }

    companion object {
        private lateinit var server: AnalysisServer
        private var result: AnalysisContext? = null
        @BeforeAll
        @Throws(Exception::class)
        @JvmStatic
        fun startup() {
            val classLoader = AnalysisServerQueriesTest::class.java.classLoader
            var resource = classLoader.getResource("good/Bouncycastle.java")
            assertNotNull(resource)

            val javaFile = File(resource.file)
            assertNotNull(javaFile)

            resource = classLoader.getResource("mark/PoC_MS1/Botan_AutoSeededRNG.mark")
            assertNotNull(resource)

            val markPoC1 = File(resource.file)
            assertNotNull(markPoC1)

            val markModelFiles = markPoC1.parent

            // Start an analysis server
            val codyze = CodyzeConfiguration()
            codyze.mark = arrayOf(File(markModelFiles))

            val config = Configuration(codyze, CpgConfiguration())
            config.executionMode.isCli = false
            config.executionMode.isLsp = false

            server = AnalysisServer(config)
            server.start()

            // Start the analysis
            val translationManager = newAnalysisRun(javaFile)
            val analyze = server.analyze(translationManager)
            try {
                result = analyze[5, TimeUnit.MINUTES]
            } catch (t: TimeoutException) {
                analyze.cancel(true)
                throw t
            }
        }

        @AfterAll
        fun teardown() {
            // Stop the analysis server
            server.stop()
        }
    }
}
